In [2]:
import pandas as pd
import numpy as np
import torch as th
from utils import DataNormalizer

In [2]:
raw_data = np.load("/home/yutong/data/air_2018.npy")
select_index = np.array([True for _ in range(raw_data.shape[1])])
select_index[[671,675]]=False
raw_data = raw_data[:,select_index,:]
raw_data.shape

(8760, 1083, 16)

In [3]:
dm = DataNormalizer()
dm.fit(raw_data[:,:,0])
dm.norm(raw_data[:,:,0])

array([[            nan, -1.76253140e-01,             nan, ...,
                    nan,             nan,  3.15913501e-01],
       [            nan, -1.76253140e-01,             nan, ...,
                    nan,             nan,  3.53236229e-01],
       [            nan,             nan,             nan, ...,
                    nan,             nan,  3.53236229e-01],
       ...,
       [-6.52004375e-03, -1.37205337e-01,  2.96651847e-01, ...,
                    nan,             nan,  8.91880229e+00],
       [ 1.07642902e-01, -1.37205337e-01,  2.35556450e-01, ...,
                    nan,             nan,  8.91880229e+00],
       [ 1.26670060e-01,  9.70814782e-02,  3.19051237e-02, ...,
                    nan,             nan,  8.91880229e+00]])

In [5]:
np.apply_along_axis(func1d=lambda x: np.nan_to_num(
    dm.fit_norm(x), nan=0), axis=0, arr=raw_data).shape

# np.nan_to_num(dm.norm(raw_data[:, :,0]), nan=0).shape

(8760, 1083, 16)

In [15]:
def test_func(data):
    print(data.shape)
    std = np.nanstd(data, axis=0, ddof=1)
    mean = np.nanmean(data, axis=0)
    return np.nan_to_num((data - mean) / std, nan=0)
    
    
np.apply_along_axis(test_func, axis=2, arr=raw_data).shape

(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,)
(16,

KeyboardInterrupt: 

In [22]:
from tqdm import tqdm
temp = np.stack([np.nan_to_num(
    dm.fit_norm(raw_data[:, :, i]), nan=0)for i in tqdm(range(raw_data.shape[2]))]).transpose(1, 2, 0)
temp.shape

100%|██████████| 16/16 [00:11<00:00,  1.36it/s]


(8760, 1083, 16)

In [18]:
temp.shape

(8760, 1083, 16)

In [36]:
pd.to_datetime(
    range(8000), unit="h", origin="2018-01-01 00:00:00")[np.random.randint(0, 2, size=8000).astype(np.bool_)].shape

(3970,)

In [38]:
pd.DataFrame(raw_data[:,:,12])

0     1     2     3     4     5     6     7     8     9     ...  1073  \
0      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   1.0   
1      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   1.0   
2      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   1.0   
3      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   1.0   
4      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   1.0   
...    ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...   ...   
8755   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0  ...   2.0   
8756   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0  ...   2.0   
8757   2.0   2.0   2.0   2.0   2.0   2.0   2.0   2.0   2.0   2.0  ...   2.0   
8758   2.0   2.0   2.0   2.0   2.0   2.0   2.0   2.0   2.0   2.0  ...   2.0   
8759   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0  ...   2.0   

      1074  1075  1076  1077  1078  1079  1080  1081  1082  
0      1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0  
1      1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0  
2      1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0  
3      1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0  
4      1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0  
...    ...   ...   ...   ...   ...   ...   ...   ...   ...  
8755   2.0   2.0   2.0   2.0   2.0   2.0   2.0   2.0   2.0  
8756   2.0   2.0   2.0   2.0   2.0   2.0   2.0   2.0   2.0  
8757   2.0   2.0   2.0   2.0   2.0   2.0   2.0   2.0   2.0  
8758   2.0   2.0   2.0   2.0   2.0   2.0   2.0   2.0   2.0  
8759   2.0   2.0   2.0   2.0   2.0   2.0   2.0   2.0   2.0  

[8760 rows x 1083 columns]

In [42]:
sample_data[np.arange(sample_data.shape[0])[:, None, None], ngb_index, :][:, :,:,0]

[]

In [9]:
th.nn.Embedding(10,16).forward(th.tensor(10)).shape

IndexError: index out of range in self

\begin{table*}[ht]
\label{exp result}
\centering
\footnotesize
\renewcommand{\arraystretch}{1.2}
\setlength{\tabcolsep}{5pt}
\caption{Model comparison on the nationwide dataset. The parameter count, denoted as \#Param, is in the order of million (M). The symbol $\Delta$ represents the reduction in MAE compared to GraphMAE. The mask ratio represents the proportion of unobserved nodes to all nodes.}
\resizebox{\textwidth}{!}{
\begin{tabular}{c|c|l|cccc|cccc|cccc}
\shline
\multirow{2}{*}{\textbf{Model}} &\multirow{2}{*}{\textbf{Year}} & \multirow{2}{*}{\textbf{Recovery Gain}} & \multicolumn{4}{c|}{\textbf{Mask Ratio = 25\%}} & \multicolumn{4}{c|}{\textbf{Mask Ratio = 50\%}} & \multicolumn{4}{c}{\textbf{Mask Ratio = 75\%}} \\
\cline{4-15}
& & & \textbf{MAE} & \textbf{$\boldsymbol{\Delta}$} & \textbf{RMSE} & \textbf{MAPE} & \textbf{MAE} & \textbf{$\boldsymbol{\Delta}$} & \textbf{RMSE} & \textbf{MAPE} & \textbf{MAE} & \textbf{$\boldsymbol{\Delta}$} & \textbf{RMSE} & \textbf{MAPE} \\
\hline
KNN &1967 & \multicolumn{1}{c|}{\multirow{2}{*}{\begin{tabular}[c]{@{}c@{}}-\\- \\ \\ \end{tabular}}}
&30.50  & +146.0\% &65.40  &1.36  &30.25  &+145.5\%  &72.23  &0.71  & 34.07 & +194.0\% & 74.55 & 0.64 \\
RF&2001 & \multicolumn{1}{c|}{} &29.22  & +135.6\% &68.95  &0.76  &29.71  & +141.2\% &71.61  &0.75  & 29.82 & +157.3\% & 70.99 & 0.74 \\
MCAM &2021 & \multicolumn{1}{c|}{0.408} & 23.94 & +93.1\% & 36.25 & 0.95 & 25.01 & +103.0\% & 37.94 & 0.92 & 25.19 & +117.3\% & 37.82 & 1.04 \\
SGNP &2019 & \multicolumn{1}{c|}{\multirow{2}{*}{\begin{tabular}[c]{@{}c@{}}0.114\\0.108 \\ \\ \end{tabular}}}
&23.60 &+90.3\% &37.58 &0.83 & 24.06 & +95.3\% & 37.08 & 0.93 & 21.68 & +87.1\% & 33.68 & 0.84 \\
STGNP &2022 & & 23.21 & +87.2\% & 38.13 & 0.62 & 21.95 & +78.2\% & 37.13 & 0.67 & 19.58 & +68.9\% & 31.95 & 0.69 \\
VAE &2013 & \multicolumn{1}{c|}{\multirow{2}{*}{\begin{tabular}[c]{@{}c@{}}0.011\\0.073 \\0.073 \\ \end{tabular}}}
& 28.49 & +129.8\% & 67.11 & 0.94 & 28.92 & +134.7\% & 69.67 & 0.94 & 29.00 & +150.2\% & 69.11 & 0.93 \\
GAE &2016 & & 12.63 & +1.9\% & 23.80 & 0.46 & 12.78 & +3.7\% & 24.11 & 0.46 & 12.57 & +8.5\% & 23.73 & 0.46 \\
GraphMAE&2022 & & 12.40 & - & 23.20 & 0.46 & 12.32 & - & 23.11 & 0.46 & 11.59 & - & 21.51 & 0.43 \\
\rowcolor{gray!20} AirNeVF&2024 & \multicolumn{1}{c|}{???} & \textbf{07.54} & \textbf{-39.2\%} & \textbf{12.70} & \textbf{0.17} & \textbf{???} & \textbf{???} & \textbf{???} & \textbf{???} & \textbf{???} & \textbf{???} & \textbf{???} & \textbf{???} \\
\shline
\end{tabular}
}
\end{table*}

In [56]:
import torch as th
s_weight = th.rand(1047, 25, 6)
s_weight = th.einsum("blc,bl->blc", s_weight,
                     s_weight.sum(dim=-1).reciprocal())
t_weight = th.cumprod(th.tensor([[0.5]]).repeat(25, 6), 0)
th.einsum("blc,lc->blc", s_weight,
          t_weight).reshape(1047,-1)

tensor([[1.6753e-01, 1.1674e-01, 2.1806e-02,  ..., 3.6644e-09, 6.3718e-09,
         6.8565e-09],
        [5.1154e-02, 1.0304e-01, 1.2273e-01,  ..., 1.3037e-09, 3.8010e-10,
         4.9247e-09],
        [1.0503e-02, 8.4317e-02, 1.3156e-01,  ..., 8.3279e-10, 1.2091e-08,
         9.8050e-10],
        ...,
        [2.8868e-02, 1.3269e-01, 2.9316e-02,  ..., 7.1694e-10, 3.9911e-09,
         5.7833e-09],
        [5.4819e-03, 1.5943e-01, 1.1731e-01,  ..., 5.5239e-11, 7.6105e-09,
         9.1828e-09],
        [1.1077e-01, 4.7571e-02, 1.0912e-01,  ..., 2.5565e-09, 6.2183e-09,
         5.8444e-09]])

In [57]:
th.tensor([0]).to("cuda",requi)

True